File: py/Sort/quickSort.py
def _quickSortPartition(nums, low, high):
    tmp = nums[low]
    while low < high:
        while low < high and nums[high] >= tmp:
            high -= 1
        nums[low] = nums[high]
        while low < high and nums[low] <= tmp:
            low += 1
        nums[high] = nums[low]
    nums[low] = tmp
    return low

def _quickSort(nums, low, high):
    if low < high:
        div_pos = _quickSortPartition(nums, low, high)
        nums = _quickSort(nums, low, div_pos-1)
        nums = _quickSort(nums, div_pos+1, high)
    return nums

def quickSort(nums, asc):
    nums = _quickSort(nums, 0, len(nums)-1)
    if not asc:
        nums = nums[::-1]
    return nums

File: cpp/Sort/quickSort.cpp
#include "Sort.h"

int Sort::quickSort(vector<int> &nums, bool asc) {
	Sort::_quickSort(nums, 0, nums.size() - 1);
	if (!asc) Sort::reverse(nums);
	return 0;
}

int Sort::_quickSort(vector<int> &nums, int low, int high) {
	if (low < high) {
		int div_pos = Sort::_quickSortPartition(nums, low, high);
		Sort::_quickSort(nums, low, div_pos - 1);
		Sort::_quickSort(nums, div_pos + 1, high);
	}
	return 0;
}

int Sort::_quickSortPartition(vector<int> &nums, int low, int high) {
	int tmp = nums[low];
	while (low < high) {
		while (low < high && nums[high] >= tmp) high--;
		nums[low] = nums[high];
		while (low < high && nums[low] <= tmp) low++;
		nums[high] = nums[low];
	}
	nums[low] = tmp;
	return low;
}
